### PR TITLE
[FIX] web_editor, web_unsplash: fix debounced search from media dialog

### PR DESCRIPTION
**Pull request by bso-odoo**
Following this flow on the Media Dialog:
- Type "hell"
- Wait a few ms so that the RPC to search images starts
- Type "o"
- The RPC of before finishes, the o is removed and no new search is done

Before this commit, the search input value of the SearchMedia component
was coming from a "needle" props. This needle prop was a state on the
parent component (FileSelector), set through a debounced handler.

This is not a good design:
- Type "hell": after 1000ms, the debounced handler will be executed with
the value "hell"
- While the handler is executing: type "o"
- When the handler finishes, it has set the needle state value to "hell"
on the parent. This will rerender the SearchMedia with "hell" as a
needle prop, and the "hello" input will be replaced with "hell".

Instead of doing that, the SearchMedia component should have its own
input state, which models the input element.
On that state, we use an effect to call the debounced "search" callback.
This way, the SearchMedia is less dependent on its parent: it only
rerenders when its input element changes.

Additionally, the fetch results are only used if they are the ones from
the last call to `search`.

Dedicated `KeepLast` instances are used for attachments, media library
and unsplash.

task-3060679

Co-authored-by: Arthur Detroux <ard@odoo.com>
Co-authored-by: Benoit Socias <bso@odoo.com>

### CODE REVIEW

File: addons/web_editor/static/src/components/media_dialog/search_media.js
Comment (rdeodoo) [inline addons/web_editor/static/src/components/media_dialog/search_media.js:29]: About the `t-model`, see review on previous PR: https://github.com/odoo/odoo/pull/104590#discussion_r1009181172

**Review by rdeodoo — APPROVED** [review]
LGTM functionally

@detrouxdev Would you have a look here?
Once it's reviewed and ok, you can either merge it yourself or I'll do it, as you prefer.

@robodoo delegate=detrouxdev

File: addons/web_editor/static/src/components/media_dialog/file_selector.xml
Comment (detrouxdev) [inline addons/web_editor/static/src/components/media_dialog/file_selector.xml:6]: as I said IRL, the removal of this needle made me think if it was possible to have a state where the FileSelector or IconSelector display results of one thing, but the input displays something else (since the two will not be synced).

And I believe it's possible in this case:
- debounce 1 (car) triggers RPC 1 which takes 5 seconds.
- debounce 2 (desks) triggers RPC 2 which take 1 second.
- RPC 2 finished and displays the result of desks
- RPC 1 which takes 5 seconds resolves, which then replaces the search results with car, while the input still displays desks

Though maybe I'm wrong, from what I've seen, debounce does not cancel / ignores promises. Instead, we should use [KeepLast](https://github.com/odoo/odoo/blob/7c9f883dc508a7a8a45bf7bf7e900da0be9b34be/addons/web/static/src/core/utils/concurrency.js#L7) as it seems to be used for other search fields which trigger RPC requests in Odoo.

Comment (bso-odoo) [reply addons/web_editor/static/src/components/media_dialog/file_selector.xml]: Added `KeepLast` mechanism and removed comments about removing the `needle` later because it is used on `loadMore` too. Now the `needle` always reflects the last `search`, be it completed or not.

**Comment by detrouxdev** [comment]
LGTM
@robodoo r+

**Comment by bso-odoo** [comment]
Staging failed on a knowledge test which relies on an overlooked feature: the ability to open the media dialog with a predefined search string.
With the failing version of this PR, when using that feature, the filtering of the search was applied, but the search field was not populated with the search string.
Because of this, the knowledge test did replace an empty search string by an empty search string - which did not trigger the unfiltered search expected by the test.
Now the behavior has been re-introduced and the test can replace an actual string by an empty search string - which therefore does trigger the expected unfiltered search and give access to the already uploaded attachment.

**Comment by bso-odoo** [comment]
Latest changes: introducing a dedicated `KeepLast` for library media and for unsplash, and returning all `keepLast.add(...)` to make sure the sequence remains the same in stable.
The three fetches could actually be made parallel in master - maybe in a distinct task after being discussed.

**Comment by detrouxdev** [comment]
@rdeodoo Looks good to me now, but since the branch has changed a bit, I thought it might be worth it if you gave the final say 😄

File: addons/web_unsplash/static/src/components/media_dialog/image_selector.js
Comment (rdeodoo) [inline addons/web_unsplash/static/src/components/media_dialog/image_selector.js:151]: could remove the `()` around `{records, isMaxed}`, no?
And somehow could've split the line so it's no that long

**Review by rdeodoo — APPROVED** [review]
@detrouxdev Sure, but don't worry in such cases you can have the final say if you are pretty confident it's developed and working as intended.

LGTM, code-wise and testing-wise.
Note: easier to reproduce the commit message bug with "Fast 3G" in chrome debug tools.

**Comment by rdeodoo** [comment]
@robodoo r+

**Pull request closed without merging**
